codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteDistributor.java

```diff
@@ -0,0 +1,428 @@
+                childState.allocate(bytesForChild);
+                bytesAllocated += bytesForChild;
+                connectionWindow -= bytesForChild;
+            }
+            int childBytesAllocated = allocateBytesForTree(child, connectionWindow);
+            bytesAllocated += childBytesAllocated;
+            connectionWindow -= childBytesAllocated;
+            return true;
+        }
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class PriorityState {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        int unallocatedStreamableBytesForTree;
+
+        PriorityState(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        /**
+         * Recursively increments the {@link #unallocatedStreamableBytesForTree()} for this branch in
+         * the priority tree starting at the current node.
+         */
+        void unallocatedStreamableBytesForTreeChanged(int delta) {
+            unallocatedStreamableBytesForTree += delta;
+            if (!stream.isRoot()) {
+                state(stream.parent()).unallocatedStreamableBytesForTreeChanged(delta);
+            }
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored before the allocated bytes are
+            // actually written.
+            unallocatedStreamableBytesForTreeChanged(-bytes);
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocate(-allocated);
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unallocatedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unallocate all bytes.
+            resetAllocated();
+
+            // Clear the streamable bytes.
+            updateStreamableBytes(0, false);
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unallocatedStreamableBytes() {
+            return streamableBytes - allocated;
+        }
+
+        int unallocatedStreamableBytesForTree() {
+            return unallocatedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class WriteVisitor implements Http2StreamVisitor {
```

Comment:
Not exactly sure what you mean by `completely wrap a Http2StreamVisitor`. In both of our suggestions, we implement `Http2StreamVisitor` ... I'm not doing much beyond that.

I'm not sure that we can ignore anything that this class is doing.  Here is the state that I'm maintaining:

1. The `Writer` ... no different from your suggestion, really.
2. The error ... I suggest we still need to re-throw at the end if we catch something. Just logging seems insufficient.
3. A `writing` boolean to prevent reentry, which we need since this code is not reentrant.